modules/core/src/lib/attribute.js

```diff
@@ -337,23 +337,35 @@ export default class Attribute extends BaseAttribute {
       opts = Object.assign({constant: false}, buffer);
     }
 
-    if (opts.value) {
-      const ArrayType = glArrayFromType(this.type || GL.FLOAT);
-      if (!(opts.value instanceof ArrayType)) {
-        log.warn(`Attribute prop ${this.id} is casted to ${ArrayType.name}`)();
-        // Cast to proper type
-        opts.value = new ArrayType(opts.value);
-      }
-    }
+    this._checkExternalBuffer(opts);
 
     this.update(opts);
+
     state.needsRedraw = true;
 
-    this._updateShaderAttributes();
     return true;
   }
 
   // PRIVATE HELPER METHODS
+  _checkExternalBuffer(opts) {
+    if (!opts.constant && opts.value) {
+      const ArrayType = glArrayFromType(this.defaultType);
+      if (
+        opts.value.BYTES_PER_ELEMENT !== ArrayType.BYTES_PER_ELEMENT &&
+        this.hasShaderAttributes
```

Comment:
Nit: this.hasShaderAttributes is the primary condition, right, so place it first in the `&&`?